### ISSUE TITLE
Article: Resist the urge to optimize

### ISSUE THREAD

**Issue by arnemertz** [Zuehlke/fifty-shades#101]
Especially in languages that are supposed to be very performant, many developers tend to feel the need to write performant code. Code gets written in a specific way "because it's faster", usually coming with a number of problems:

* Readability is sacrificed to the performance god and clean code principles are willfully or ignorantly violated. (reference to #59 )
* It is not proven by a profiler that the code, written that way, is actually more performant. Instead, it is "known" from war stories and lore from the past that this is The Way Things Have to be Done :tm: (Reference to #31 )
* It is not proven by a profiler that performance is actually _needed_ at that point in the code. 
* While focusing on the low-level optimizations one may oversee that the overarching algorithm is not as effective as it could be.
* Optimizers are actually prevented from doing their work because the code does not demonstrate well-known patterns that are recognized by the optimizer.

Bottom line: 

1. Write clean and maintainable code first
2. Check your algorithms for efficiency
3. Use benchmarking and profiling to show that and where optimizations are needed
3. Only as a last resort apply manual optimizations, guided and verified by the profiler

**Comment by abeggchr** [comment]
Are you still working at Zühlke? I can't find you in the address book...

**Comment by arnemertz** [comment]
I am, at the Hamburg office.
